timeutil/schedule.go

```diff
@@ -150,34 +154,159 @@ func findNthWeekDay(t time.Time, weekday time.Weekday, nthInMonth uint) time.Tim
+	anchoredAtStart := ws.AnchoredAtStart()
+
+	if t.After(endDay) || t.Before(startDay) {
+		// outside of dates range of the week span, consider edge cases:
+		// - next month if the span is anchored at the end (eg. mon-fri1 30.07-03.08, t=31.07)
+		// - previous month if the span is anchored at the start (eg. fri4-thu 27.07-02.08, t=01.08)
+
+		if anchoredAtStart {
+			// eg. fri4-thu, range anchored at previous month
+			if matchingWeekdaysInMonth(t) != 1 {
+				// no match if t is not within the first week
 				return false
 			}
+			prevMonth := monthPrev(t)
+			startDay, endDay = ws.dateRangeAnchoredAt(prevMonth)
 		} else {
-			startDay := findNthWeekDay(t, start.Weekday, start.Pos)
-			endDay := findNthWeekDay(t, end.Weekday, end.Pos)
-
-			if t.Day() < startDay.Day() || t.Day() > endDay.Day() {
+			// eg. mon-fri1, range anchored at the next month
+			if !isLastWeekdayInMonth(t) {
+				// no match if t is not within the last week
 				return false
 			}
-			return true
+			nextMonth := monthNext(t)
+			startDay, endDay = ws.dateRangeAnchoredAt(nextMonth)
+		}
+		// check whether t matches the range that spills from the
+		// previous month or from the next month
+		if t.After(endDay) || t.Before(startDay) {
+			return false
 		}
 	}
+	return weekdayMatch(t)
+}
 
-	if wdStart <= wdEnd {
-		// single day (mon) or start < end (eg. mon-fri)
-		return t.Weekday() >= wdStart && t.Weekday() <= wdEnd
+// monthNext returns the first day of the next month relative to t
+func monthNext(t time.Time) time.Time {
+	n := t
+	// advance by 28 days at most, so that we don't skip a 28 day February
+	n = n.AddDate(0, 0, 28)
+	for n.Month() == t.Month() {
+		n = n.Add(24 * time.Hour)
 	}
-	// wraps around the week end, eg. fri-mon
-	return t.Weekday() >= wdStart || t.Weekday() <= wdEnd
+	if n.Day() != 1 {
+		// backtrack if we didn't land on the first day yet
+		n = n.AddDate(0, 0, -n.Day()+1)
+	}
+	return n
+}
+
+// monthPrev returns the last day of previous month relative to t
+func monthPrev(t time.Time) time.Time {
+	return t.AddDate(0, 0, -1*(t.Day()+1))
+}
+
+// AnchoredAtStart returns true when the week span is anchored at the starting
+// point, or false otherwise
+func (ws WeekSpan) AnchoredAtStart() bool {
+	return ws.Start.Pos != EveryWeek
+}
+
+// dateRangeAnchoredAt returns the range of dates that match the week span, with the
+// anchor sharing the same month as t
+func (ws WeekSpan) dateRangeAnchoredAt(t time.Time) (start, end time.Time) {
+	weekPos := ws.Start.Pos
+	anchoredAtStart := ws.AnchoredAtStart()
+	if !anchoredAtStart {
+		weekPos = ws.End.Pos
+	}
+	// first check the start/end dates in the same month as t
+	if weekPos != LastWeek {
+		start = findNthWeekDay(t, ws.Start.Weekday, weekPos)
+		end = findNthWeekDay(t, ws.End.Weekday, weekPos)
+	} else {
+		start = findLastWeekDay(t, ws.Start.Weekday)
+		end = findLastWeekDay(t, ws.End.Weekday)
+	}
+
+	if start.After(end) || (start.Equal(end) && !ws.IsSingleDay()) {
```

Comment:
under what situation could the second part of the `||` be true? by my naive reading of this, if start == end then by definition the span has to be a single day and `ws.IsSingleDay()` will be true as well. I assume there's some subtlety in the above block assigning to start and end but I can't seem to think of a case that would have a multi-day span and somehow result in start and end being the same day.

Reply:
Eg. `mon1-mon` is an 8 day span that falls into this scenario. Actually 8 days is the maximum span we can define, since `mon1-tue` is interpreted as from first Monday to the following Tuesday (2 days).

Now, after adding a comment and a test case (since it's quite subtle indeed), I noticed a bug elsewhere, which got fixed in one of the patches I pushed.

Reply:
/me wonders if we should allow `mon1-mon`...